File: src/main/java/ankol/mod/merger/core/BaseModManager.java
package ankol.mod.merger.core;

import ankol.mod.merger.core.filetrees.PathFileTree;
import ankol.mod.merger.tools.ColorPrinter;
import ankol.mod.merger.tools.Localizations;
import ankol.mod.merger.tools.Tools;
import cn.hutool.cache.Cache;
import cn.hutool.cache.CacheUtil;
import cn.hutool.core.io.IoUtil;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
import org.apache.commons.compress.archivers.zip.ZipFile;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.function.Function;

/**
 * 基准MOD分析器 - 负责加载和分析基准MOD（原版文件）
 *
 * @author Ankol
 */
@Slf4j
public class BaseModManager {

    /**
     * 基准MOD文件路径
     */
    private final Path baseModPath;

    /**
     * 文件名 → 标准路径的映射
     * 键：文件名（小写）
     * 值：在基准MOD中的相对路径
     */
    @Getter
    private Map<String, PathFileTree> indexedBaseModFileMap;

    /**
     * 基准MOD是否已加载
     */
    @Getter
    private boolean loaded = false;

    /**
     * 临时文件缓存目录
     */
    private final Path cacheDir;

    /**
     * 已提取文件的缓存映射：相对路径 → 临时文件路径
     */
    private final Map<String, Path> extractedFileCache;
    @SuppressWarnings("rawtypes")
    private final Cache<String, ParsedResult> BASE_TREE_CACHE = CacheUtil.newFIFOCache(100, 30 * 1000);

    /**
     * 构造函数
     *
     * @param baseModPath 基准MOD文件路径
     */
    public BaseModManager(Path baseModPath) {
        this.baseModPath = baseModPath;
        this.indexedBaseModFileMap = new LinkedHashMap<>();
        this.extractedFileCache = new LinkedHashMap<>();
        // 创建缓存目录：temp/BaseModCache_时间戳
        this.cacheDir = Path.of(Tools.getTempDir(), "BaseModCache_" + System.currentTimeMillis());
        try {
            Files.createDirectories(this.cacheDir);
        } catch (IOException e) {
            ColorPrinter.warning("Failed to create base mod cache directory: " + e.getMessage());
        }
    }

    /**
     * 加载基准MOD（优化：只读取条目，不解压文件）
     *
     * @throws IOException 如果基准MOD文件不存在或无法读取
     */
    public void load() throws IOException {
        if (loaded) {
            ColorPrinter.warning(Localizations.t("BASE_MOD_ALREADY_LOADED"));
            return;
        }
        if (!Files.exists(baseModPath)) {
            throw new IOException(Localizations.t("BASE_MOD_FILE_NOT_FOUND", baseModPath));
        }

        try {
            long startTime = System.currentTimeMillis();
            this.indexedBaseModFileMap = Tools.indexPakFile(baseModPath); //这里构建的索引MAP里还没有真正解压出来文件
            loaded = true;
            long elapsed = System.currentTimeMillis() - startTime;
            ColorPrinter.success(Localizations.t("BASE_MOD_INDEXED_FILES",
                    indexedBaseModFileMap.size(),
                    baseModPath.getFileName(),
                    elapsed
            ));
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * 从基准MOD中提取指定文件的内容（带缓存优化）
     *
     * @param relPath 文件在基准MOD中的相对路径
     * @return 文件内容，如果文件不存在返回null
     */
    public String extractFileContent(String relPath) throws IOException {
        if (!loaded) {
            return null;
        }

        // 规范化路径（统一使用小写文件名查找）
        String fileName = Tools.getEntryFileName(relPath).toLowerCase();
        PathFileTree pathFileTree = indexedBaseModFileMap.get(fileName);
        if (pathFileTree == null) {
            return null;
        }

        String fileEntryName = pathFileTree.getFileEntryName();
        Path cachedFile = extractedFileCache.get(fileEntryName);
        if (cachedFile != null && Files.exists(cachedFile)) {
            return Files.readString(cachedFile);
        }

        String content = extractFileFromPak(baseModPath, fileEntryName);
        if (content != null) {
            try {
                String safeFileName = fileEntryName.replace("/", "_").replace("\\", "_");
                Path tempFile = cacheDir.resolve(safeFileName);
                Files.createDirectories(tempFile.getParent());
                Files.writeString(tempFile, content);
                extractedFileCache.put(fileEntryName, tempFile);
            } catch (IOException e) {
                ColorPrinter.warning("Failed to cache extracted file: " + fileEntryName + " - " + e.getMessage());
            }
        }
        return content;
    }

    /**
     * 判断MOD里的文件路径是否正确
     *
     * @param filePath mod文件路径
     */
    public boolean hasPathConflict(String filePath) {
        if (!loaded) {
            return false;
        }
        String fileName = Tools.getEntryFileName(filePath);
        PathFileTree pathFileTree = indexedBaseModFileMap.get(fileName);
        //有时会有一些不属于mod的文件被加入到pak中，这里查到空后说明不是原版mod支持修改的文件.
        if (pathFileTree == null) {
            return false;
        }
        String correctPath = pathFileTree.getFileEntryName();
        return correctPath != null && !correctPath.equalsIgnoreCase(filePath);
    }

    /**
     * 获取建议的修正路径
     *
     * @param filePath 待检查的文件相对路径
     * @return 如果存在同名文件，返回基准MOD中的正确路径；否则返回null
     */
    public String getSuggestedPath(String filePath) {
        if (!loaded) {
            return null;
        }
        String fileName = Tools.getEntryFileName(filePath);
        return indexedBaseModFileMap.get(fileName).getFileEntryName();
    }

    /**
     * 清理临时文件缓存
     * 建议在合并完成后调用此方法释放磁盘空间
     */
    public void clearCache() {
        if (!Files.exists(cacheDir)) {
            return;
        }

        try (var stream = Files.walk(cacheDir)) {
            stream.sorted(java.util.Comparator.reverseOrder())
                    .forEach(path -> {
                        try {
                            Files.delete(path);
                        } catch (IOException e) {
                            // 忽略删除错误
                        }
                    });
            extractedFileCache.clear();
        } catch (IOException e) {
            ColorPrinter.warning("Failed to clear base mod cache: " + e.getMessage());
        }
    }

    /**
     * 从基准MOD获得解析后的语法树节点，带缓存机制
     *
     * @param fileEntryName 文件在压缩包中的全路径
     * @param function      解析语法树使用的函数
     * @return 解析结果，如果文件不存在返回null
     */
    @SuppressWarnings("unchecked")
    public <T extends BaseTreeNode> ParsedResult<T> parseForm(String fileEntryName, Function<String, ParsedResult<T>> function) {
        return BASE_TREE_CACHE.get(fileEntryName,
                () -> {
                    String content = extractFileContent(fileEntryName);
                    if (content == null) {
                        return null;
                    }
                    return function.apply(content);
                });
    }

    /**
     * 从PAK文件中提取指定文件的内容
     *
     * @param pakFile       PAK文件
     * @param fileEntryName 文件在PAK中的相对路径
     * @return 文件内容，如果文件不存在返回null
     */
    public String extractFileFromPak(Path pakFile, String fileEntryName) throws IOException {
        try (ZipFile zipFile = ZipFile.builder().setPath(pakFile).get()) {
            ZipArchiveEntry entry = zipFile.getEntry(fileEntryName);
            if (entry.getSize() == 0) {
                return null;
            }

            try (var inputStream = zipFile.getInputStream(entry)) {
                return IoUtil.read(inputStream, StandardCharsets.UTF_8);
            }
        }
    }
}

File: src/main/java/ankol/mod/merger/core/BaseModManager.kt
package ankol.mod.merger.core

import ankol.mod.merger.core.filetrees.BaseFile
import ankol.mod.merger.domain.ParsedResult
import ankol.mod.merger.exception.ExitProcessException
import ankol.mod.merger.tools.ConsoleColorPrinter
import ankol.mod.merger.tools.Localizations.t
import ankol.mod.merger.tools.SoftLruCache
import ankol.mod.merger.tools.Tools.getEntryFileName
import ankol.mod.merger.tools.logger
import net.sf.sevenzipjbinding.IInArchive
import net.sf.sevenzipjbinding.SevenZip
import net.sf.sevenzipjbinding.impl.RandomAccessFileInStream
import net.sf.sevenzipjbinding.simple.ISimpleInArchiveItem
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.RandomAccessFile
import java.lang.AutoCloseable
import java.nio.file.Path
import kotlin.io.path.exists

/**
 * Base MOD Manager
 * Responsible for operations related to the base MOD
 *
 * @author Ankol
 */
class BaseModManager(private val basePakDirPath: Path) : AutoCloseable {
    private val log = logger()

    /**
     * Mapping from file name to standard path
     * Key: file name (lowercase)
     * Value: relative path in the base MOD
     */
    var indexedBaseModFileMap: MutableMap<String, BaseFile> = mutableMapOf()

    /**
     * Whether the base MOD has been loaded
     */
    var isLoaded = false

    private val basePakSet = setOf("data0.pak", "data1.pak", "databt.mpak")

    /** AST tree cache **/
    private val astTreeCache = SoftLruCache<String, ParsedResult<*>>(2048)

    /** File content cache **/
    private val fileContentCache = SoftLruCache<String, CachedBaseFile>(2048)

    private data class CachedBaseFile(
        val content: String,
    )

    /**
     * Reuse zip-connection to avoid repeatedly opening and closing the same file, which is costly.
     * Keep it open until the merge is complete, then close it in the close() method.
     */
    private lateinit var zipFileConnections: Map<String, IInArchive>

    init {
        load()
    }

    /**
     * Load the base MOD
     */
    fun load() {
        if (isLoaded) {
            ConsoleColorPrinter.warning(t("BASE_MOD_ALREADY_LOADED"))
            return
        }
        val startTime = System.currentTimeMillis()
        this.zipFileConnections = openArchiveConnection()
        this.indexedBaseModFileMap = indexBasePack(zipFileConnections)
        this.isLoaded = true
        val timetake = System.currentTimeMillis() - startTime
        ConsoleColorPrinter.success(t("BASE_MOD_INDEXED_FILES", basePakDirPath.fileName, indexedBaseModFileMap.size, timetake))
    }

    /**
     * Validate the base MOD path and establish Zip connections
     */
    private fun openArchiveConnection(): MutableMap<String, IInArchive> {
        if (!basePakDirPath.exists()) {
            throw IOException(t("BASE_MOD_FILE_NOT_FOUND", basePakDirPath))
        }
        val connectionMap = mutableMapOf<String, IInArchive>()
        for (pakName in basePakSet) {
            try {
                val resolvePath = basePakDirPath.resolve(pakName)
                if (resolvePath.exists()) {
                    //Establish archive connection
                    val inArchive = SevenZip.openInArchive(
                        null,
                        RandomAccessFileInStream(RandomAccessFile(resolvePath.toFile(), "r"))
                    )
                    connectionMap[pakName] = inArchive
                }
            } catch (e: RuntimeException) {
                log.error("Failed to create zip file $pakName. Reason: ${e.message}", e)
            }
        }
        if (connectionMap.isEmpty()) {
            throw ExitProcessException(1, t("BASE_MOD_NO_VALID_PAK_FOUND", basePakDirPath))
        }
        return connectionMap
    }

    /**
     * Index the base game PAK, create a mapping table for the files inside for subsequent operations
     */
    private fun indexBasePack(archiveConnection: Map<String, IInArchive>): MutableMap<String, BaseFile> {
        val pakIndexMap = HashMap<String, BaseFile>()
        archiveConnection.forEach { (pakName, inArchive) ->
            inArchive.simpleInterface.archiveItems.forEach { archiveEntry: ISimpleInArchiveItem ->
                if (!archiveEntry.isFolder) {
                    val archiveEntryName = archiveEntry.path
                    val fileName = getEntryFileName(archiveEntryName)
                    if (fileName in pakIndexMap) {
                        ConsoleColorPrinter.warning(
                            t(
                                "TOOLS_SAME_FILE_NAME_WARNING",
                                fileName,
                                archiveEntryName,
                                pakIndexMap[fileName]?.archiveEntryName
                            )
                        )
                    }
                    pakIndexMap[fileName] = BaseFile(fileName, archiveEntryName, archiveEntry.itemIndex, pakName)
                }
            }
        }
        return pakIndexMap
    }

    /**
     * Extract the content of the specified file from the base MOD (with cache optimization)
     *
     * @param entryFilePath Relative path of the file in the base MOD
     * @return File content, or null if the file does not exist
     */
    fun extractFileContent(entryFilePath: String): String? {
        if (!isLoaded) {
            return null
        }
        val fileName = getEntryFileName(entryFilePath)
        val baseFile = indexedBaseModFileMap[fileName] ?: return null

        val fileEntryName = baseFile.archiveEntryName
        val cached = fileContentCache.get(fileEntryName)
        if (cached != null) {
            return cached.content.ifEmpty { null }
        }

        val extracted = extractFileFromPak(baseFile.basePakName, baseFile)
        fileContentCache.put(fileEntryName, extracted)
        if (extracted.content.isEmpty()) {
            return null
        }
        return extracted.content
    }

    /**
     * Extract the specified file from the PAK file and cache the content and hash in memory
     */
    private fun extractFileFromPak(basePakName: String, entryBaseFile: BaseFile): CachedBaseFile {
        val inArchive = zipFileConnections[basePakName] ?: throw IOException(t("BASE_MOD_FILE_NOT_FOUND", basePakName))
        var content = ""
        ByteArrayOutputStream().use { os ->
            inArchive.extractSlow(entryBaseFile.archiveEntryIndex) { data ->
                os.write(data)
                data.size
            }
            content = os.toString(Charsets.UTF_8)
        }

        return CachedBaseFile(content)
    }

    /**
     * Check whether the file path in the MOD is correct
     *
     * @param fileEntryName MOD file path
     */
    fun hasPathConflict(fileEntryName: String): Boolean {
        if (!isLoaded) {
            return false
        }
        val fileName = getEntryFileName(fileEntryName)
        val pathFileTree = indexedBaseModFileMap[fileName] ?: return false
        // Sometimes files not belonging to the mod are added to the pak. If found empty, it's not a file supported by the original mod.
        val correctPath = pathFileTree.archiveEntryName
        return !correctPath.equals(fileEntryName, ignoreCase = true)
    }

    /**
     * Get the suggested corrected path
     *
     * @param filePath Relative path of the file to check
     * @return If a file with the same name exists, return the correct path in the base MOD; otherwise return null
     */
    fun getSuggestedPath(filePath: String): String? {
        if (!isLoaded) {
            return null
        }
        val fileName = getEntryFileName(filePath)
        return indexedBaseModFileMap[fileName]?.archiveEntryName
    }

    /**
     * Get the parsed syntax tree node from the base MOD, with cache mechanism
     *
     * @param fileEntryName Full path of the file in the archive
     * @param parseFunction Function used to parse the syntax tree
     * @return Parse result, or null if the file does not exist
     */
    @Suppress("UNCHECKED_CAST")
    fun <T : BaseTreeNode> parseForm(
        fileEntryName: String,
        parseFunction: (String) -> ParsedResult<T>
    ): ParsedResult<T>? {
        val entryFileName = getEntryFileName(fileEntryName)
        val canonicalEntryName = indexedBaseModFileMap[entryFileName]?.archiveEntryName ?: fileEntryName

        val cached = astTreeCache.get(canonicalEntryName)
        if (cached != null) {
            return cached as ParsedResult<T>
        }
        val content = extractFileContent(canonicalEntryName) ?: return null
        val result = parseFunction(content)
        astTreeCache.put(canonicalEntryName, result)
        return result
    }

    /**
     * Clear memory cache and close ZipFile connections
     * It is recommended to call this method to release resources after merging is complete
     */
    override fun close() {
        // Close ZipFile connections
        try {
            zipFileConnections.forEach { (_, zipFile) -> zipFile.close() }
        } catch (e: IOException) {
            ConsoleColorPrinter.warning("Failed to close ZipFile connection: " + e.message)
        }
        fileContentCache.clear()
        astTreeCache.clear()
    }
}
